overlord/devicestate/devicemgr.go

```diff
@@ -141,6 +143,52 @@ func Manager(s *state.State, hookManager *hookstate.HookManager, runner *state.T
 	return m, nil
 }
 
+// StartUp implements StateStarterUp.Startup.
+func (m *DeviceManager) StartUp() error {
+	m.state.Lock()
+	defer m.state.Unlock()
+
+	if release.OnClassic {
+		return nil
+	}
+
+	mode := m.SystemMode()
+	if mode != "run" && mode != "recover" {
+		return nil
+	}
+
+	var seeded bool
+	err := m.state.Get("seeded", &seeded)
+	if err != nil && err != state.ErrNoState {
+		return err
+	}
+	if !seeded {
+		return nil
+	}
+
+	// get the disk that we mounted the ubuntu-seed partition from
+	disk, err := disks.DiskFromMountPoint(boot.InitramfsUbuntuSeedDir, nil)
+	if err != nil {
+		return err
+	}
+	partUUID, err := disk.FindMatchingPartitionUUID("ubuntu-data-enc")
+	if err != nil {
+		// no encrypted partitions, don't retrieve the auth key
+		logger.Noticef("no encrypted partitions found")
+		return nil
+	}
+	encdev := filepath.Join("/dev/disk/by-partuuid", partUUID)
+
+	k, err := secboot.AuthKeyFromKernelKeyring(encdev)
```

Comment:
we need to decide what to do if we are in the case that the recovery.key was used and not the actual sealed objects... we can't just have snapd die on that

Reply:
This is an interesting situation. We could load the value we saved beforehand but it feels like defeating the purpose of having the auth key passed through the kernel keyring. But if we say that in the case of unlocking with the recovery key the user should also explicitly reseal also assumes that the user has interaction capabilities to do so (such as a terminal) which doesn't seem to be always the case. So, in the case of the PCR unlock failing on a secure system without a user interface, should the system heal itself automatically? This may have security implications too.